Hazelcast.Net/Hazelcast.Client.Spi/ProxyManager.cs

```diff
@@ -106,105 +104,137 @@ public void Destroy()
+                throw ExceptionUtil.Rethrow(e);
             }
+        }
+
+        private ICollection<IDistributedObject> GetLocalDistributedObjects()
+        {
+            return new ReadOnlyCollection<IDistributedObject>(_proxies.Values.ToList<IDistributedObject>());
+        }
 
+        private ClientProxy InitProxyLocal(string service, string id, Type requestedInterface)
+        {
+            var proxy = MakeProxy(service, id, requestedInterface);
             proxy.SetContext(new ClientContext(_client.GetSerializationService(), _client.GetClientClusterService(),
-                _client.GetClientPartitionService(), _client.GetInvocationService(), _client.GetClientExecutionService(),
+                _client.GetClientPartitionService(), _client.GetInvocationService(),
+                _client.GetClientExecutionService(),
                 _client.GetListenerService(), _client.GetNearCacheManager(), this, _client.GetClientConfig()));
             proxy.PostInit();
+            return proxy;
+        }
 
-            _proxies.AddOrUpdate(ns, n => proxy, (n, oldProxy) => {
-                Logger.Warning("Replacing old proxy for " + oldProxy.GetName() + " of type " + oldProxy.GetType() + " with " + proxy.GetType());
+        public ClientProxy GetOrCreateProxy<T>(string service, string id) where T : IDistributedObject
+        {
+            var objectInfo = new DistributedObjectInfo(service, id);
+            var requestedInterface = typeof(T);
+            var clientProxy = _proxies.GetOrAdd(objectInfo, distributedObjectInfo =>
+            {
+                // create a new proxy, which needs initialization on server.
+                var proxy = InitProxyLocal(service, id, requestedInterface);
+                InitializeWithRetry(proxy);
                 return proxy;
             });
-            return proxy;
+
+            // only return the existing proxy, if the requested type args match
+            var proxyInterface = clientProxy.GetType().GetInterface(requestedInterface.Name);
+            var proxyArgs = proxyInterface.GetGenericArguments();
+            var requestedArgs = requestedInterface.GetGenericArguments();
+            if (proxyArgs.SequenceEqual(requestedArgs))
+            {
+                // the proxy we found matches what we were looking for
+                return clientProxy;
+            }
+
+            var isAssignable = true;
+            for (int i = 0; i < proxyArgs.Length; i++)
+            {
+                if (!proxyArgs[i].IsAssignableFrom(requestedArgs[i]))
+                {
+                    isAssignable = false;
+                    break;
+                }
+            }
+
+            if (isAssignable)
+            {
+                _proxies.TryRemove(objectInfo, out clientProxy);
+                return GetOrCreateProxy<T>(service, id);
+            }
+
+            throw new InvalidCastException(string.Format("Distributed object is already created with {0}", proxyArgs));
```

Comment:
This gives the impression that the problem is to re-create an existing proxy whereas the problem is the incompatibility of the types. I would say `...already created with incompatible types {0}`